### PR TITLE
feat(web): Studio Builder PR-3 — connected mode

### PR DESCRIPTION
**Pull request by seanrobertwright**
## Summary

- **Problem:** The PR-2 builder could only edit throwaway fixtures — no way to load, save, create, rename, or delete a real workflow from the console UI.
- **Why it matters:** Authors can now build and maintain workflow YAML visually instead of hand-editing files, completing M1 of the Archon Studio integration.
- **What changed:** A connected route (`/console/builder[/:name]`) that loads/saves real workflows via the existing CRUD endpoints, a project picker, explicit Save with a dirty + nav guard, full create/rename/delete, server-tier validation surfaced into the issue panel, and bundled→Save-as.
- **What did NOT change (scope boundary):** No server changes (reuses existing `GET`/`PUT`/`DELETE /api/workflows/:name`, `POST /api/workflows/validate`, `GET /api/commands` untouched), no DB, no other package, no production `/workflows/builder` route. `BuilderPage` stays a controlled component — its only change is an additive `extraIssues` prop.

### Before

```
User                    Builder (/console/builder)
────                    ──────────────────────────
opens builder ────────▶ seeds BuilderPage from a FIXTURE (switcher)
edits nodes ──────────▶ in-memory only; header says "load/save lands next milestone"
(no way to persist) ──x  nothing reaches disk
```

### After

```
User                    BuilderConnected (/console/builder[/:name])
────                    ───────────────────────────────────────────
picks project ────────▶ [resolves cwd; persisted + ?project= deep-link]
opens workflow ───────▶ [loadWorkflow(name,cwd) -> fromWorkflowDefinition]
edits nodes ──────────▶ dirty dot; import+client issues in IssueList
clicks Save ──────────▶ [client-validate -> validateWorkflow -> PUT -> invalidate]
                        [server 400 detail surfaces as a source:'server' issue]
New / Rename / Delete ▶ [create-on-save / new-then-old rename / guarded delete]
bundled workflow ─────▶ [read-only banner; "Save as" writes a project override]
navigates away dirty ─▶ [confirm prompt]; reload dirty -> [beforeunload prompt]
```

### After

```
ConsoleApp routes
  └─[~] /console/builder[/:name] ── [+] BuilderConnected.tsx
                                       ├── useBuilderProject (localStorage + ?project=)
                                       ├── skills/workflows: load/save/delete/validate  ──> existing /api/workflows*
                                       ├── [+] connect/save-logic.ts (pure: issues, planRename, name validation)
                                       ├── store/cache useEntity + invalidate ; [~] store/keys K.workflow(cwd,name)
                                       └─[~] BuilderPage (additive extraIssues prop) ── IssueList
  [-] BuilderRoute.tsx (fixture route deleted; fixture switcher already lives in PreviewPage)
```

**Connection inventory:**

| From | To | Status | Notes |
|------|----|--------|-------|
| ConsoleApp | BuilderConnected | **new** | mounts `builder` + `builder/:name` |
| BuilderConnected | skills/workflows | **new** | load/save/delete/validate verbs |
| skills/workflows | /api/workflows* | unchanged | existing endpoints, no server change |
| BuilderConnected | connect/save-logic | **new** | pure issue/rename/name logic |
| BuilderConnected | BuilderPage | **modified** | additive `extraIssues` prop |
| ConsoleApp | BuilderRoute | **removed** | fixture route deleted |

## Label Snapshot

- Risk: `risk: low`
- Size: `size: M`
- Scope: `web`
- Module: `web:console-builder`

## Change Metadata

- Change type: `feature`
- Primary scope: `web`

## Linked Issue

- Closes # — n/a (one of several PRs; tracking issue closes when the series lands)
- Related # coleam00/Archon#1863 (Archon Studio integration; M1 / PR-3)
- Depends on # PR-2 (coleam00/Archon#2015) — **merged** into `dev`. This branch is rebased onto `upstream/dev`, so the PR contains only the single PR-3 commit.

## Validation Evidence (required)

```bash
bun run type-check        # exit 0 (all 10 packages)
bun run lint              # exit 0 (zero warnings)
bun run format:check      # exit 0
bun --filter @archon/web test   # 338 pass / 0 fail (incl. 19 new)
bun --filter @archon/web build  # exit 0
```

- Evidence: re-validated after rebasing onto `upstream/dev` (`59bbd00b`) — type-check (0), format (0), web tests 338 pass / 0 fail, and the PR-3/console files lint clean. All gates green. `bun run validate`'s only failure is a pre-existing, environment-specific Windows test in `@archon/providers` (`pathKind > broken symlink` → `EPERM`; symlink creation needs Developer Mode/admin). That package is untouched by this PR and the test passes on Linux CI.
- Skipped: Level-4 manual steps (no DOM/browser test harness — a console house rule); the nav-guard UX and visual surface are manual gates.

## Security Impact (required)

- New permissions/capabilities? `No`
- New external network calls? `No` (reuses existing workflow endpoints)
- Secrets/tokens handling changed? `No`
- File system access scope changed? `No` (writes go through the existing server-validated `PUT`, which validates `cwd` against registered codebases)

## Compatibility / Migration

- Backward compatible? `Yes`
- Config/env changes? `No`
- Database migration needed? `No`

## Human Verification (required)

- Verified scenarios: type-check/lint/format/build/tests green; code-reviewed (1 HIGH race + 2 LOW items found and fixed, re-validated).
- Edge cases checked (unit + reasoning): server 400 → `source:'server'` issue; blocking client errors gate Save; rename collision/no-op/invalid-name blocked; rename delete-fail → non-fatal warning; bundled Save-as; create-on-save flips to edit mode; controlled-select revert on cancelled nav-guard; `beforeunload` teardown.
- What was not verified: live-server Level-4 manual walkthrough (no DOM harness).

## Side Effects / Blast Radius (required)

- Affected subsystems: `packages/web/src/experiments/console/` only.
- Potential unintended effects: none outside the console builder; saving normalizes YAML key order (lossless but not byte-identical — documented).
- Guardrails: isolation ESLint guard, no-logging rule, 19 unit tests.

## Rollback Plan (required)

- Fast rollback: revert this single commit (`af0ba60b`); nothing outside the console experiment changes.
- Feature flags: none (behind the `/console/*` experiment surface).
- Observable failure symptoms: builder fails to load/save a workflow; issue panel shows a `source:'server'` error.

## Risks and Mitigations

- Risk: nav guard does not intercept the browser Back button or external `ProjectRail` clicks (the app is a non-data `<BrowserRouter>`, so `useBlocker` is unavailable).
  - Mitigation: `beforeunload` + confirm-on-intercept on the builder's own controls; documented as a known limitation; data-router migration is out of scope.
- Risk: subfoldered workflows don't load via the single-name GET.
  - Mitigation: surfaced as a clear "not found → offer New" empty state; documented; matches PR-2's limitation.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Launched connected workflow builder routes at **/console/builder** and **/console/builder/:name** with deep-link editing.
  * Added project picker with persisted selection (URL `?project=` sync) and full CRUD: **New, Save, Save as (bundled read-only), Rename, Delete**.
* **Bug Fixes**
  * Improved workflow save validation and server error reporting, including merged/deduplicated issue display and stricter workflow-name checks.
  * Enhanced rename behavior to handle collisions/no-ops more reliably.
* **Tests**
  * Added unit tests for save/rename logic and workflow URL building.
* **Documentation**
  * Updated builder and console README plus added builder terminology glossary.

### CODE REVIEW

**Comment by seanrobertwright** [comment]
## PR Review Summary — multi-agent (7 agents)

7 specialized agents reviewed the diff (`upstream/dev...HEAD`, 14 files). Advisory only — no files modified. Suite is green (**522 pass / 0 fail**); type-check, format, and the console-subtree lint all pass.

### Critical Issues (0)
None. No correctness-blocking or security issues; isolation guard, no-`console.*` rule, and explicit-return-type rules are all clean.

### Important Issues (2 — recommend fixing before merge)
| Agent | Issue | Location |
|-------|-------|----------|
| silent-failure-hunter · type-design | `validateWorkflow` returning `{ valid:false }` with **empty/absent `errors`** calls `setServerIssues([])` → panel silently clears and Save/Rename re-enables with no explanation (infinite "why won't it save?" loop). Latent today (the server always populates `errors`), but the type permits it. Fix: emit a fallback issue when `errors` is empty **and** tighten `ValidateWorkflowResponse` to a discriminated union (`{valid:true} \| {valid:false; errors: string[]}`) to drop the `?? []`. | `BuilderConnected.tsx:259` (doSave) & `:328` (doRename); `skills/workflows.ts:111` |
| silent-failure-hunter | **All** `loadWorkflow` errors (500 / 403 / network drop — not just 404) render the "No workflow named X was found … Create a new workflow" state. On a transient server error for a workflow that exists, this misleads and the recovery action invites a **duplicate**. Fix: gate `notFound` on `loadView.error instanceof HttpError && status === 404`; show a distinct "failed to load" state otherwise. | `BuilderConnected.tsx:444`, `:578` |

### Suggestions (nice to have)
| Agent | Suggestion | Location |
|-------|------------|----------|
| type-design | `WorkflowSource` is declared twice (silent-divergence risk) — re-export from `primitives/workflow.ts` instead. | `skills/workflows.ts:92` |
| silent-failure · comment | Rename-delete warning and the `serverErrorToIssues` empty-body fallback both say "**Save** failed" / leak raw `HttpError.message` (internal URL + JSON) for a DELETE. Reuse `serverErrorToIssues` for the detail; use a verb-neutral fallback. | `BuilderConnected.tsx:337`; `save-logic.ts:20` |
| silent-failure · code-reviewer | `doDelete` lacks `finally { setBusy(false) }`. **Not a current bug** (success path unmounts the component), but cheap hygiene against a future same-component navigation. | `BuilderConnected.tsx:289` |
| pr-test-analyzer | Extract the two pure functions still inside the component (`errorToIssues`, `renameReasonMessage`) into `save-logic.ts` so they're unit-testable; add tests for the empty-`error` body branch and `planRename` precedence ordering. | `BuilderConnected.tsx:68,80` |
| code-simplifier | Extract a `clientIssue(rule,message)` helper for the 4 near-identical instant-error blocks; collapse the redundant `loadKey`/`loadView` guards; reuse `editorKey` for the `BuilderPage` `key` (currently duplicated with a `name` vs `name ?? ''` divergence); inline single-use `pid`; truthiness check in `serverErrorToIssues` (~−18 lines). | `BuilderConnected.tsx:141,182,233,245…`; `save-logic.ts:23` |
| type-design | `RenamePlan.steps` is unused (decorative) — drive execution from it or drop it; annotate `extraIssues` as `readonly Issue[]`. | `save-logic.ts:85`; `BuilderPage.tsx:56` |
| comment-analyzer | `openOptions` comment examples are no-ops (real cases are create-mode/timing); "≤200-char truncated" is actually ≤203 (`...` suffix) in two places; drop/inline the 4 `Spike #N` refs (they point into gitignored `.agents/plans/`). | `BuilderConnected.tsx:432`; `lib/http.ts:21`; `save-logic.ts:15` |
| code-reviewer | Consolidate the split `listWorkflows` import into the existing block (cosmetic; below threshold). | `BuilderConnected.tsx:51` |

### Documentation Issues
| Agent | Issue | Location |
|-------|-------|----------|
| docs-impact | `PR-2 (this layer)` is stale now that PR-2 is merged → `PR-2 (merged)`. | `builder/README.md:2` |
| docs-impact | Status section still says "In progress: the `builder/` subtree" though PR-1→PR-3 are now shipped. | `README.md` (Status) |

(Confirmed correct: the new `archon.console.builderProject` localStorage key, the `/console/builder[/:name]` routes, and the full PR-3 specifics are all already documented.)

### Strengths
- **code-reviewer: PASS** — isolation guard, zero `console.*`, explicit return types all clean; the editorKey-ref reset guard is sound against the post-save stale-while-revalidate refetch; rename new-then-old atomicity is correct.
- **type-design**: `IssueId`/`makeIssue` rated 9.75/10 (textbook branded type); `RenamePlan` 8.5; `HttpError.bodySnippet` 8; `WorkflowSaveSource` correctly encodes "bundled is never a write target."
- **comment-analyzer: GOOD** — contract claims verified true: `isValidWorkflowName` is an exact parity mirror of the server's `isValidCommandName`, `validateWorkflow` genuinely returns HTTP 200 on invalid, and the subdir limitation is real.
- **pr-test-analyzer**: extraction discipline is principled — every pure transform in `save-logic.ts` is tested at the right boundary.
- **silent-failure-hunter**: 5 patterns confirmed safe (localStorage guards, `res.text().catch`, `void doX()` handlers, rename-warning surfacing).

### Verdict
**NEEDS FIXES** — no blockers, but the two Important items (silent blocked-save on empty `errors`; load errors masquerading as "not found") are worth addressing before merge. Everything else is advisory polish.

### Recommended Actions
1. Add the empty-`errors` fallback + make `ValidateWorkflowResponse` a discriminated union (kills the silent-block in `doSave`/`doRename`).
2. Distinguish 404 from other `loadWorkflow` errors; show a real error state instead of "not found → Create".
3. Optional cleanups: re-export `WorkflowSource`; extract `errorToIssues`/`renameReasonMessage` (+ tests); `clientIssue` helper; doc nits.

<sub>🤖 Generated via /prp-review-agents (code-reviewer · docs-impact · pr-test-analyzer · comment-analyzer · silent-failure-hunter · type-design-analyzer · code-simplifier)</sub>

File: packages/web/src/experiments/console/builder/BuilderConnected.tsx [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+    try {
+      await deleteWorkflow(name, { cwd, source: saveTargetFor(effectiveSource) });
+      invalidate(K.workflows(cwd));
+      invalidate(K.workflow(cwd, name));
+      navigate(`/console/builder${projectQuery}`);
+    } catch (e) {
+      setServerIssues(errorToIssues(e, 'delete.failed', 'Delete failed (unknown error).'));
+      setBusy(false);
+    }
+  }, [name, cwd, effectiveSource, navigate, projectQuery]);
```
[the remark this answers is not in the document]
Comment (seanrobertwright) [reply packages/web/src/experiments/console/builder/BuilderConnected.tsx:277]: Fixed in 583e587f — `doDelete` now clears `busy` in a `finally` block (previously only on the catch path). Independent review noted the success path navigates `builder/:name` → `builder`; because that route reuses a single `BuilderConnected` instance rather than remounting, the unconditional `finally` reset is the safe fix. Thanks!

File: packages/web/src/experiments/console/builder/BuilderConnected.tsx [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+    });
+  }, [cwd, existingNames, navigate, projectQuery]);
+
+  // --- Navigation controls (dirty-guarded) ---------------------------------
+  const onPickProject = useCallback(
+    (id: string): void => {
+      confirmIfDirty(() => {
+        setProjectId(id);
+        navigate(`/console/builder?project=${encodeURIComponent(id)}`);
+      });
```
[the remark this answers is not in the document]
Comment (seanrobertwright) [reply packages/web/src/experiments/console/builder/BuilderConnected.tsx:377]: Fixed in 583e587f — `onPickProject` now normalizes the empty "Select a project…" option to `undefined` and omits `?project=` when there's no selection. I also hardened the hook: `useBuilderProject` treats a bare/empty `?project=` as no-selection (`searchParams.get('project') || undefined`), so an empty param can't shadow the persisted default on reload.

File: packages/web/src/experiments/console/builder/BuilderConnected.tsx [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+  const openOptions = useMemo(() => {
+    const names = new Set(existingNames);
+    if (name !== undefined) names.add(name);
+    return [...names].sort((a, b) => a.localeCompare(b));
+  }, [existingNames, name]);
+
+  const saveLabel = readOnly ? 'Save as' : 'Save';
+  const canSave = !busy && (readOnly || dirty || isCreateMode);
+  const workflowOpen = name !== undefined && imported !== null;
+  const notFound = name !== undefined && !isCreateMode && loadView.error !== undefined;
```
[the remark this answers is not in the document]
Comment (seanrobertwright) [reply packages/web/src/experiments/console/builder/BuilderConnected.tsx]: Fixed in 583e587f — the load error is now split: `notFound = loadError instanceof HttpError && loadError.status === 404` (still offers "Create a new workflow"), while a new `workflowLoadError` branch renders 500/403/network failures as "Failed to load …" instead of the misleading not-found state. This matched an independent multi-agent finding, so double-confirmed. Thanks for catching it.

File: packages/web/src/experiments/console/builder/connect/use-builder-project.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  // Seed once: the deep-link param wins over the persisted default. Later URL
+  // changes are driven by `setProjectId` callers, not re-read here.
+  const [projectId, setProjectIdState] = useState<string | undefined>(
+    () => paramProject ?? readStored()
+  );
+
+  const setProjectId = useCallback((id: string | undefined): void => {
+    setProjectIdState(id);
+    writeStored(id);
+  }, []);
```
[the remark this answers is not in the document]
Comment (seanrobertwright) [reply packages/web/src/experiments/console/builder/connect/use-builder-project.ts:59]: Fixed in 583e587f — added a `useEffect` in `useBuilderProject` that follows a later `?project=` change (guarded with `paramProject !== projectId` to avoid a write-back loop against `BuilderConnected`'s URL-sync effect). Browser back/forward and cross-project builder links now track the URL instead of leaving `projectId` stale.

File: packages/web/src/experiments/console/store/keys.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ export const K = {
   projects: 'projects' as const,
   project: (id: string): string => `project:${id}`,
   workflows: (cwd: string): string => `workflows:${cwd}`,
+  workflow: (cwd: string, name: string): string => `workflow:${cwd}:${name}`,
```
[the remark this answers is not in the document]
Comment (seanrobertwright) [reply packages/web/src/experiments/console/store/keys.ts]: Fixed in 583e587f — `K.workflow` now `encodeURIComponent`s both `cwd` and `name`, so a `:` in either value can no longer collapse distinct `(cwd, name)` pairs onto the same cache entry. Producer and consumer both build the key via `K.workflow`, so they stay aligned.

**Comment by seanrobertwright** [comment]
## Follow-up: multi-agent review — actions taken

All findings from the 7-agent review above have been addressed in **583e587f** (on top of **835f5966**). The two Important items and every actionable suggestion were applied; one simplification was intentionally skipped (justified below). Gates re-run green: repo-root `type-check` (0), `format` (0), console `eslint --max-warnings 0` (0), web tests **352 pass / 0 fail** (+14 new).

### Important (both applied)
| Finding | Action |
|---|---|
| `validateWorkflow` `valid:false` with empty `errors` → silent panel clear / re-enabled Save | **Applied** — new pure `validationFailureToIssues()` guarantees a fallback issue when `errors` is empty/absent; `ValidateWorkflowResponse` is now a discriminated union (`{valid:true} \| {valid:false; errors?}`). Used in both `doSave` and `doRename`. |
| All `loadWorkflow` errors render "not found → Create" | **Applied** — `notFound` now gates on `HttpError.status === 404`; non-404 failures render a distinct `workflowLoadError` state. (Same as CodeRabbit's Major finding.) |

### Suggestions (applied)
| Finding | Action |
|---|---|
| `WorkflowSource` duplicated (silent-divergence risk) | **Applied** — re-exported from `primitives/workflow` (single source of truth). |
| Extract `errorToIssues` / `renameReasonMessage` for testability | **Applied** — moved both to `save-logic.ts`; also extracted `clientIssue`, `errorDetail`, `validationFailureToIssues`. |
| Add empty-`error`-body + `planRename` precedence tests | **Applied** — plus new tests for the extracted helpers (+14 unit tests). |
| `clientIssue()` helper for the 4 instant-error blocks | **Applied.** |
| Reuse `editorKey` for the `BuilderPage` key; inline single-use `pid`; truthiness in `serverErrorToIssues` | **Applied.** |
| `RenamePlan.steps` unused (false promise) | **Applied** — dropped the field; success variant is now `{ ok: true }` (tests updated). |
| `readonly extraIssues` | **Applied.** |
| `doDelete` missing `finally` | **Applied** — code-reviewer judged it a non-bug (navigation unmounts), but it's cheap hardening and CodeRabbit flagged it Major, so `finally { setBusy(false) }` it is. |
| Rename-delete warning / `serverErrorToIssues` "Save failed" fallback leaked raw `HttpError.message` and was verb-specific | **Applied** — warning now uses `errorDetail()` (parsed server message); fallback is verb-neutral ("Request failed"). |
| Comment nits: `openOptions` example, "≤200/203-char" precision (×2), editorKey null-branch, drop gitignored `Spike #N` refs | **Applied** (all). |
| Consolidate split `listWorkflows` import | **Applied.** |
| Docs: `builder/README` "PR-2 (this layer)"→"(merged)"; `console/README` "In progress" Status | **Applied.** |

### Not applied (justified)
| Finding | Why not |
|---|---|
| code-simplifier: collapse the redundant `loadKey` / `loadView` guards using non-null assertions (`cwd!`, `name!`) | **Skipped.** `@typescript-eslint/no-non-null-assertion` is `error` in this repo, so the assertions won't lint. The repeated `cwd === undefined \|\| name === undefined` checks are load-bearing for TypeScript's control-flow narrowing (a plain `idle` boolean can't narrow the other vars), so the verbose form stays. |

Thanks to the review agents — the silent-block and load-error items were the highest-value catches.

<sub>🤖 review-response generated with Claude Code</sub>
